repair/repair.cc

```diff
@@ -1966,23 +1988,57 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get0();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring ks_source_dc;
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (!source_dc_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    ssize_t rf = nt_strat.get_replication_factor(source_dc);
+                    // Can we ensure a quorum in source_dc?
+                    if (full_racks_in_source_dc.size() > rf / 2) {
```

Comment:
I think this is incorrect. Nothing prevents a user to have a single rack inside a DC.  Say we have2 dcs, each have 3 nodes in the cluster.  All nodes are alive and good. This patch will prevent replace or rebuild from the same DC.

Reply:
I can give another example why this full racks check is incorrect. Say a DC with rack r1,r2,r3,r4,r5. RF =3. each rack has 2 nodes. Say r3 r4 r5 are down. We sill have full_racks_in_souce_dc =2  >  RF / 2.  This allows rebuild/replace from the source dc, but 6 out of 10 nodes in the DC are down. The decision to allow using source_dc is wrong.

Reply:
new algorithm implemented in v4